### PR TITLE
Lower MatrixMapCols

### PR DESCRIPTION
**Pull request by tpoterba**
* redesign LowerMatrixIR to pass through mutable state for relational lets
 * switch several functions on Python frontend to define functions; this
   is to manage the size of generated code

### CODE REVIEW

**Comment by tpoterba** [comment]
This is a big PR, but the bulk of the changed lines come from indentation changes. I'll annotate various pieces to help the review process along. I'll also think about how to break it into smaller pieces.

File: hail/src/main/scala/is/hail/expr/ir/ForwardRelationalLets.scala [hunk 1]
```diff
@@ -22,8 +22,9 @@ object ForwardRelationalLets {
         case x@RelationalRef(name, _) =>
           val (n, nd) = usages(name)
           usages(name) = (n + 1, math.max(nd, nestingDepth.lookup(x)))
-        case _ => ir1.children.foreach(visit)
+        case _ =>
       }
+      ir1.children.foreach(visit)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/ForwardRelationalLets.scala]: was just wrong before - wouldn't recur into children of relational let nodes.

File: hail/src/main/scala/is/hail/expr/ir/Exists.scala [hunk 2]
```diff
@@ -31,16 +31,54 @@ object Exists {
...
+    case AggExplode(_, _, _, isScan) => !isScan
+    case AggGroupBy(_, _, isScan) => !isScan
+    case AggArrayPerElement(_, _, _, _, isScan) => !isScan
+    case _ => false
+  }
 }
 
 object ContainsAgg {
-  def apply(root: BaseIR): Boolean = Exists(root, _.isInstanceOf[ApplyAggOp])
+  def apply(root: IR): Boolean = IsAggResult(root) || (root match {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/Exists.scala]: The previous implementations were wrong in that they would recur into the value children of things like TableAggregate, even if that wasn't part of the same scope.

File: hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala [hunk 3]
```diff
@@ -6,49 +6,41 @@ import is.hail.expr.types.virtual.{TArray, TInt32, TInterval}
 import is.hail.utils._
 
 object LowerMatrixIR {
-  val entriesFieldName = MatrixType.entriesIdentifier
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala]: I'm not sure why the diff is so terribly horrible for this file. My changes involved:

 * switching from using partial functions to match statements with catch-all base cases. There is no base case for MatrixIR 🎉 
 * doing the typecheck after these match statements
 * Adding the lowering rule for MatrixMapCols

That's it.

**Comment by tpoterba** [comment]
Timings:

Master:
```
running sample_qc...
    run 1 took 27.46s
    run 2 took 24.08s
    run 3 took 23.59s
    Mean, Median: 25.04s, 24.08s

running variant_and_sample_qc...
    run 1 took 54.03s
    run 2 took 45.70s
    run 3 took 45.14s
    Mean, Median: 48.29s, 45.70s
```

PR:
```
running sample_qc...
    run 1 took 18.89s
    run 2 took 17.22s
    run 3 took 16.28s
    Mean, Median: 17.47s, 17.22s

running variant_and_sample_qc...
    run 1 took 38.78s
    run 2 took 39.05s
    run 3 took 35.66s
    Mean, Median: 37.83s, 38.78s
```

**Comment by tpoterba** [comment]
I removed a couple of unnecessary changes to be added later. This is conceptually small enough for review, even though the diff is still oddly large dude to the LowerMatrixIR change.

File: hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala [hunk 4]
```diff
@@ -6,49 +6,41 @@ import is.hail.expr.types.virtual.{TArray, TInt32, TInterval}
...
+    ab.result().foldRight[MatrixIR](l1) { case ((ident, value), body) => RelationalLetMatrixTable(ident, value, body) }
   }
 
-  private[this] def lowerChildren(ir: BaseIR): BaseIR = {
-    val loweredChildren = ir.children.map(lower)
+
+  private[this] def lowerChildren(ir: BaseIR, ab: ArrayBuilder[(String, IR)]): BaseIR = {
+    val loweredChildren = ir.children.map {
+      case tir: TableIR => lower(tir, ab)
+      case mir: MatrixIR => lower(mir, ab)
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala]: very nice.

File: hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala [hunk 5]
```diff
@@ -92,408 +84,522 @@ object LowerMatrixIR {
...
-          } else
-            b0
+        var b0 = lift(newCol)
+        val aggs = aggBuilder.result()
+        val scans = scanBuilder.result()
+
+        val idx = Ref(genUID(), TInt32())
+        val idxSym = Symbol(idx.name)
+
+        val aggTransformer: IRProxy => IRProxy = if (aggs.nonEmpty) {
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala]: definitely.

**Comment by tpoterba** [comment]
d'oh -- there's still no lowering rule for `MatrixMultiWrite`. Will PR that separately.

**Comment by tpoterba** [comment]
MatrixMultiWrite lowering is in.

File: hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala [hunk 6]
```diff
@@ -92,410 +72,535 @@ object LowerMatrixIR {
...
-      }
 
-      val lc = lower(child)
-      val e = Env[IR]("va" -> Ref("row", lc.typ.rowType),
-        "global" -> SelectFields(Ref("global", lc.typ.globalType), child.typ.globalType.fieldNames))
-      lc.mapRows(
-        liftScans(Subst(newRow, BindingEnv(e, scan = Some(e), agg = Some(e))))
-          .insertFields(entriesField -> 'row (entriesField)))
-    }
+        val scanTransformer: IRProxy => IRProxy = if (scans.nonEmpty) {
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala]: This should flip to `if (scans.isEmpty)` to match `aggTransformer`.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala]: Addressed. Also added a test for a rebase collision bug I found.

**Pull request merged**